content/courses/intro-to-embedded-sys-prog/chapters/low_level_programming.rst

```diff
@@ -1,38 +1,408 @@
+code to interact with it. That way the bulk of the code |mdash| the
+usage |mdash| is like any other Ada code, and doesn't propagate the
+low-level details all over the client code. This greatly simplifies
+usage and understandability as well as easing porting to new
+hardware-specific aspects. You change things in one place, rather than
+everywhere.
+
+For example, consider a device mapped to the memory address space of the
+processor. To interact with the device we interact with one or more
+memory cells. Reading input from the device amounts to reading the value
+at the associated memory location. Likewise, sending output to the
+device amounts to writing to that location.
+
+To represent this device mapping we declare a variable of an appropriate
+type and specify the starting address the object should occupy. (There
+are other ways too, but for a single, statically mapped object this is
+the simplest approach.) We'd want to specify some other characteristics
+as well, but let's focus on the address.
+
+.. image:: images/memory-mapped-device.png
+  :width: 600
+  :alt: Memory-mapped device with Ada object declared at specific address
+
+If the hardware presents an interface consisting of multiple fields
+within individual memory cells, we can use a record type instead of a
+single unsigned type representing a single word. Ada allows us to
+specify the exact record layout, down to the individual bit level, for
+any types we may need to use for the record components. When we
+declare the object we use that record type, again specifying the
+starting address. Then we can just refer to the object's record
+components as usual, having the compiler compute the address offsets
+required to access the components representing the individual hardware
+fields.
+
+Note that we aren't saying that other languages cannot do this too. Many
+can, using good programming practices. What we're saying is that those
+practices are designed into the Ada way of doing it.
 
 
 Guaranteed Level of Support
 ---------------------------
 
-.. todo::
+The Ada reference manual has an entire section dedicated to low-level
+programming. That's section 13, "Representation Issues," which provides
+facilities for developers to query and control aspects of various
+entities in their code, and for interfacing to hardware. Want to specify
+the exact layout for a record type's components? See section 13. Want to
+specify the alignment of a type? See section 13. And that's just the
+beginning. We'll talk about these facilities as we go, but there's
+another point to make about this section.
 
-    Complete section!
+In particular, section 13 includes recommended levels of support to be
+provided by language implementations, i.e., compilers and other
+associated tools. Although the word "recommended" is used, the
+recommendations are meant to be followed.
+
+For example, section 13.3 says that, for some entity named X, "X'Address
```

Comment:
Again, you could use a link here:

```
For example, `section 13.3 <https://www.adaic.org/resources/add_content/standards/12rm/html/RM-13-3.html>`_ says that...
```

Also, in previous course, we've been using the `:ada:` syntax for every Ada keyword, and even for variables or subprograms from a source-code example. In this case, you should write:

```
for some entity named :ada:`X`, ":ada:`X'Address` ...
```

Reply:
I don't want to point people to the RM. After the Ada 95 version it is too complicated and formal for any other than the compiler developers.